src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -176,6 +176,134 @@ protected void LookupMembersInternal(LookupResult result, NamespaceOrTypeSymbol
+            }
+
+            static NamedTypeSymbol? inferExtensionTypeArguments(NamedTypeSymbol extension, TypeSymbol receiverType, CSharpCompilation compilation, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                if (extension.Arity == 0)
+                {
+                    return extension;
+                }
+
+                var containingAssembly = extension.ContainingAssembly;
+                var conversions = containingAssembly.CorLibrary.TypeConversions;
+
+                // Note: we create a value for purpose of inferring type arguments even when the receiver type is static
+                var syntax = (CSharpSyntaxNode)CSharpSyntaxTree.Dummy.GetRoot();
+                var receiverValue = new BoundLiteral(syntax, ConstantValue.Bad, receiverType) { WasCompilerGenerated = true };
+
+                var typeArguments = MethodTypeInferrer.InferTypeArgumentsFromReceiverType(extension, receiverValue, compilation, conversions, ref useSiteInfo);
+                if (typeArguments.IsDefault || typeArguments.Any(t => !t.HasType))
+                {
+                    return null;
+                }
+
+                bool success = checkConstraints(extension, typeArguments, compilation, conversions, ref useSiteInfo);
+                if (!success)
+                {
+                    return null;
+                }
+
+                return extension.Construct(typeArguments);
+            }
+
+            static bool checkConstraints(NamedTypeSymbol symbol, ImmutableArray<TypeWithAnnotations> typeArgs, CSharpCompilation compilation,
+                TypeConversions conversions, ref CompoundUseSiteInfo<AssemblySymbol> useSiteInfo)
+            {
+                var typeParams = symbol.TypeParameters;
+                var diagnosticsBuilder = ArrayBuilder<TypeParameterDiagnosticInfo>.GetInstance();
+                var substitution = new TypeMap(typeParams, typeArgs);
+                ArrayBuilder<TypeParameterDiagnosticInfo>? useSiteDiagnosticsBuilder = null;
+
+                bool success = symbol.CheckConstraints(
+                    new ConstraintsHelper.CheckConstraintsArgs(compilation, conversions, includeNullability: false, NoLocation.Singleton, diagnostics: null, template: new CompoundUseSiteInfo<AssemblySymbol>(useSiteInfo)),
+                    substitution, typeParams, typeArgs, diagnosticsBuilder, nullabilityDiagnosticsBuilderOpt: null,
+                    ref useSiteDiagnosticsBuilder, ignoreTypeConstraintsDependentOnTypeParametersOpt: null);
+
+                diagnosticsBuilder.Free();
+
+                if (useSiteDiagnosticsBuilder != null && useSiteDiagnosticsBuilder.Count > 0)
+                {
+                    foreach (TypeParameterDiagnosticInfo diagnostic in useSiteDiagnosticsBuilder)
+                    {
+                        useSiteInfo.Add(diagnostic.UseSiteInfo);
+                    }
+                }
```

Comment:
Free `useSiteDiagnosticsBuilder`?

Reply:
No, that one isn't pooled. See `ensureUseSiteDiagnosticsBuilder` which initializes this when needed.